crates/uv/src/commands/project/sync.rs

```diff
@@ -881,3 +900,270 @@ fn store_credentials_from_target(target: InstallTarget<'_>) {
+    Replace,
+    /// Create a new environment.
+    Create,
+}
+
+/// Represents the action taken during a dry run.
+///
+/// this struct describe the simulated action applied to a
+/// resource (environment, lockfile, package)
+#[derive(Serialize, Debug)]
+#[serde(rename_all = "snake_case")]
+enum LockAction {
+    /// No changes are needed.
+    AlreadyExist,
+    /// The lock would be updated.
+    Update,
+    /// Create a new lock.
+    Create,
+}
+
+/// The dry run environment details, including path, action, and Python configuration.
+#[derive(Serialize, Debug)]
+struct SyncReport {
+    /// The path to the environment.
+    env_path: PathBuf,
+    /// The kind of environment this is.
+    kind: EnvKind,
+    /// Whether this is a dry run.
+    dry: bool,
+    /// The action taken for the environment.
+    action: SyncAction,
+    // Python executable path.
+    python_executable: PathBuf,
+    // Python full version.
+    python_version: String,
+}
+
+impl ProjectReport {
+    /// Creates a new `SyncEntry` with the given project directory.
+    fn new(project_dir: &Path) -> Self {
+        Self {
+            project_dir: project_dir.to_owned(),
+            sync: None,
+            lock: None,
+        }
+    }
+
+    /// Prints the entire report as JSON
+    fn print_json(&self, printer: Printer) -> Result<()> {
+        let output = serde_json::to_string_pretty(self)?;
+        writeln!(printer.stdout(), "{output}")?;
+        Ok(())
+    }
+
+    /// Prints the sync component in human readable form
+    fn print_sync_text(&self, printer: Printer) -> Result<()> {
+        let Some(SyncReport {
+            env_path: path,
+            kind,
+            dry,
+            action,
+            python_executable: _,
+            python_version: _,
+        }) = &self.sync
+        else {
+            return Ok(());
+        };
+
+        match (kind, action, dry) {
+            (EnvKind::Project, SyncAction::AlreadyExist, true) => {
+                writeln!(
+                    printer.stderr(),
+                    "{}",
+                    format!(
+                        "Discovered existing environment at: {}",
+                        path.user_display().bold()
+                    )
+                    .dimmed()
+                )?;
+            }
+            (EnvKind::Project, SyncAction::AlreadyExist, false) => {
+                // Currently intentionally silent
+            }
+            (EnvKind::Project, SyncAction::Replace, true) => {
+                writeln!(
+                    printer.stderr(),
+                    "{}",
+                    format!(
+                        "Would replace existing virtual environment at: {}",
+                        path.user_display().bold()
+                    )
+                    .dimmed()
+                )?;
+            }
+            (EnvKind::Project, SyncAction::Replace, false) => {
+                // Currently intentionally silent
+            }
```

Comment:
Entries like this are me trying to make explicit all the places where the old code kinda "randomly" didn't print anything. Helped make it clear to me that my refactor was correct.